orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/support/TriggerDeserializer.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.model.support
+
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.netflix.spinnaker.orca.pipeline.model.*
+
+internal class TriggerDeserializer
+  : StdDeserializer<Trigger>(Trigger::class.java) {
+  override fun deserialize(parser: JsonParser, context: DeserializationContext): Trigger =
+    parser.codec.readTree<JsonNode>(parser).run {
+      Trigger(
+        get("type").textValue(),
+        parseTriggerPayload(parser),
+        get("correlationId")?.textValue(),
+        get("user")?.textValue() ?: "[anonymous]",
+        get("parameters")?.mapValue(parser) ?: mutableMapOf(),
+        get("artifacts")?.listValue(parser) ?: mutableListOf(),
+        get("notifications")?.listValue(parser) ?: mutableListOf(),
+        get("rebake")?.booleanValue() == true,
+        get("dryRun")?.booleanValue() == true,
+        get("strategy")?.booleanValue() == true
+      )
+    }
+
+  private fun JsonNode.parseTriggerPayload(parser: JsonParser): TriggerPayload? {
+    return when {
+      looksLikeDocker() -> parseValue<DockerTriggerPayload>(parser)
+      looksLikeJenkins() -> parseValue<JenkinsTriggerPayload>(parser)
+      looksLikePipeline() -> parseValue<PipelineTriggerPayload>(parser)
+      looksLikeGit() -> parseValue<GitTriggerPayload>(parser)
+      else -> null
+    }
```

Comment:
Identifies what type of payload based on the fields that exist in the incoming JSON.

Reply:
What if it's a webhook with an arbitrary payload, does this just become null?

Reply:
How would anything in Orca reference it?

Reply:
Using SPEL

Reply:
ah, yeah. Maybe I should add a catch all here that's just a `Map<String, Object>`